### PR TITLE
[hotfix] Using Flink ExceptionUtils for exception checking

### PR DESCRIPTION
**Pull request by haoxins**
## What is the purpose of the change

The `cause instanceof FlinkJobNotFoundException` cat't catch the exception

```
... SessionJobReconciler [ERROR][.../...] Failed to cancel job fffffffffab07a8e0000000000000001, will reschedule after 5000 milliseconds.
java.util.concurrent.ExecutionException: org.apache.flink.runtime.rest.util.RestClientException: [org.apache.flink.runtime.rest.handler.RestHandlerException: Job could not be found.
	at org.apache.flink.runtime.rest.handler.job.JobCancellationHandler.lambda$handleRequest$0(JobCancellationHandler.java:128)
	...
Caused by: org.apache.flink.runtime.messages.FlinkJobNotFoundException: Could not find Flink job (fffffffffab07a8e0000000000000001)
	at org.apache.flink.runtime.dispatcher.Dispatcher.cancelJob(Dispatcher.java:749)
	...
Caused by: org.apache.flink.runtime.rest.util.RestClientException: [org.apache.flink.runtime.rest.handler.RestHandlerException: Job could not be found.
	at org.apache.flink.runtime.rest.handler.job.JobCancellationHandler.lambda$handleRequest$0(JobCancellationHandler.java:128)
	...
Caused by: org.apache.flink.runtime.messages.FlinkJobNotFoundException: Could not find Flink job (fffffffffab07a8e0000000000000001)
	at org.apache.flink.runtime.dispatcher.Dispatcher.cancelJob(Dispatcher.java:749)
	...

```

## Brief change log

Using Apache commons lang for exception cause checking

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: (no)
  - Core observer or reconciler logic that is regularly executed: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Review by mxm — APPROVED** [review]
LGTM

**Review by 1996fanrui — COMMENTED** [review]
Thanks @haoxins for the fix!

The checkstyle[1] is not passed, please run `mvn spotless:apply` and try to build on your Local first, thanks~

[1] https://github.com/apache/flink-kubernetes-operator/actions/runs/8778998199/job/24132429588?pr=818#step:5:11735

**Comment by haoxins** [comment]
> Thanks @haoxins for the fix!
> 
> The checkstyle[1] is not passed, please run `mvn spotless:apply` and try to build on your Local first, thanks~
> 
> [1] https://github.com/apache/flink-kubernetes-operator/actions/runs/8778998199/job/24132429588?pr=818#step:5:11735

done

**Review by 1996fanrui — COMMENTED** [review]
It seems 2 unit tests are failed.

https://github.com/apache/flink-kubernetes-operator/actions/runs/8794050120/job/24133050257?pr=818#step:5:27827

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java [hunk 1]
```diff
@@ -127,14 +128,14 @@ public DeleteControl cleanupInternal(FlinkResourceContext<FlinkSessionJob> ctx)
                                     : UpgradeMode.STATELESS;
                     cancelJob(ctx, upgradeMode);
                 } catch (ExecutionException e) {
-                    final var cause = e.getCause();
-
-                    if (cause instanceof FlinkJobNotFoundException) {
+                    if (ExceptionUtils.indexOfType(e, FlinkJobNotFoundException.class) != -1) {
```
Comment (1996fanrui) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java:131]: IIUC, we wanna find is there `FlinkJobNotFoundException` inside of e, right?

If so,  `org.apache.flink.util.ExceptionUtils#findThrowable(e, FlinkJobNotFoundException.class).isPresent()` may be better here. It's easy to understand to check the index.

WDYT?

Comment (1996fanrui) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java:131]: Also, would you mind adding a unit test to check whether it is expected when the sub cause is `FlinkJobNotFoundException`?

Comment (haoxins) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java:131]: > Also, would you mind adding a unit test to check whether it is expected when the sub cause is

We don't need to add an e2e or real session job UT, only need to check the exception can be caught, right?

**Review by caicancai — APPROVED** [review]
LGTM

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java (hunk 1 above)
Comment (1996fanrui) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java]: It's up to you. Both of e2e ITcase or Unit test are fine for me.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java [hunk 2]
```diff
@@ -34,6 +34,7 @@
 import org.apache.flink.kubernetes.operator.utils.StatusRecorder;
 import org.apache.flink.runtime.messages.FlinkJobNotFoundException;
 import org.apache.flink.runtime.messages.FlinkJobTerminatedWithoutCancellationException;
+import org.apache.flink.util.ExceptionUtils.findThrowable;
```
Comment (1996fanrui) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java]:
```suggestion
import static org.apache.flink.util.ExceptionUtils.findThrowable;
```

it's static import, and please run `mvn spotless:apply` after change.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java (hunk 1 above)
Comment (haoxins) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconciler.java]: I changed the mock service to follow the real behavior.

File: flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/TestingFlinkService.java [hunk 3]
```diff
@@ -447,7 +450,10 @@ private String cancelJob(FlinkVersion flinkVersion, JobID jobID, boolean savepoi
         }
 
         if (isFlinkJobNotFound) {
-            throw new FlinkJobNotFoundException(jobID);
+            throw new RestClientException(
+                    "Job could not be found.",
+                    new FlinkJobNotFoundException(jobID),
+                    HttpResponseStatus.NOT_FOUND);
```
[the remark this answers is not in the document]
Comment (haoxins) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/TestingFlinkService.java:456]: > and throw pure FlinkJobNotFoundException or FlinkJobNotFoundException as the sub cause randomly.

This sounds good to me, will change it~

File: flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/TestingFlinkService.java [hunk 4]
```diff
@@ -84,15 +87,7 @@
...
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (haoxins) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/TestingFlinkService.java:90]: but this is finished by ` mvn spotless:apply` :)

[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/TestingFlinkService.java:90]: flink has 2 checkstyles, the CI starts run for now, you can check it later.

[the remark this answers is not in the document]
Comment (haoxins) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/TestingFlinkService.java]: sry, I rerun `mvn spotless:apply`

**Review by 1996fanrui — APPROVED** [review]
Thanks for the update, LGTM

**Pull request merged**